src/sync.js

```diff
@@ -779,7 +782,13 @@ SyncApi.prototype._processSyncResponse = async function(syncToken, data) {
         client.store.storeAccountDataEvents(events);
         events.forEach(
             function(accountDataEvent) {
-                if (accountDataEvent.getType() == 'm.push_rules') {
+                // XXX: This is awful: ignore push rules from our cached sync. We fetch the
+                // push rules before syncing so we actually have up-to-date ones. We do want
+                // to honour new push rules that come down the sync but synapse doesn't
+                // put new push rules in the sync stream when the base rules change, so
+                // if the base rules change, we do need to refresh. We therefore ignore
+                // the push rules in our cached sync response.
+                if (accountDataEvent.getType() == 'm.push_rules' && !isCachedResponse) {
```

Comment:
My worry here is that the js sdk now has to worry about where the sync data is coming from (obviously).  Would the first incremental sync not update the push rules if they've changed? If so, can we instead ignore `client.pushRules` until the first incremental sync is done?

Reply:
Well, we have a convenient flag for both so either is fine with me really (`isCachedResponse` or `syncOptions.hasSyncedBefore`). The difference is essentially whether we discard the pushrules in an initial /sync. Strictly speaking it's more correct to clobber the pushrules we fetched earlier with the ones from the initial /sync because then if there was a change in the time between us fetching them and us syncing, we'll have the up to date ones.For that reason I'd probably argue for using `isCachedResponse`.

Reply:
IRL we chatted and agreed that this is probably OK for now. But the server really ought to be sending new push rules if the base rules get updated.